### PR TITLE
`<ranges>`: Fix the return types and overload resolution of range adaptor closure objects

### LINKED ISSUE #4153: `<ranges>`: `operator|(_Left&& __l, _Right&& __r)` should return `decltype(auto)`
https://github.com/microsoft/STL/blob/bc5a5c349643f35b65d755c2ce33861c8029f705/stl/inc/ranges#L183-L190

IIUC, LWG-3981 being tagged as NAD means it's a bug in MSVC-STL, right?

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4153.

Notes:
1. IIRC `_Pipeline::operator()` overloads also need to return `decltype(auto)` per [[range.adaptor.object]/1](https://eel.is/c++draft/range.adaptor.object#1).
2. Deleted overloads are needed, see #4057 and [_Perfect forwarding call wrappers need `=delete`_](https://quuxplusone.github.io/blog/2021/07/30/perfect-forwarding-call-wrapper/).
3. There seems to be a bug of MSVC on cv-qualification in constraints, which is reported as DevCom-10519736.
4. I _think_ we don't need to guard anything for C++23 (but not 100% sure currently) - in C++20 the return types should always be controlled by the implementation and should always be cv-unqualified class types.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
Thanks! I think this needs only one maintainer approval (although it's near the limits of my ranges knowledge).

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by CaseyCarter — APPROVED** [review]

**Pull request merged**
